File: problems/09_heap_priority_queue/0621_task_scheduler/solution.swift
// Problem: 0621. Task Scheduler
// Difficulty: Medium

import Foundation

class Solution {
    func leastInterval(_ tasks: [Character], _ n: Int) -> Int {
        var cnt: [Character: Int] = [:]
        for t in tasks { cnt[t, default: 0] += 1 }
        let maxF = cnt.values.max() ?? 0
        let maxCount = cnt.values.filter { $0 == maxF }.count
        let formula = (maxF - 1) * (n + 1) + maxCount
        return max(tasks.count, formula)
    }
}

func checkInt(_ a: Int, _ e: Int, _ msg: String) {
    if a != e { fatalError("FAIL \(msg): \(a) vs \(e)") }
}

let sol = Solution()
checkInt(sol.leastInterval(["A", "A", "A", "B", "B", "B"], 2), 8, "ex1")
checkInt(sol.leastInterval(["A", "A", "A", "B", "B", "B"], 0), 6, "no cool")
checkInt(sol.leastInterval(["A", "A", "A", "A", "A", "A", "B", "C", "D", "E", "F", "G"], 2), 16, "big")
checkInt(sol.leastInterval(["A"], 2), 1, "one")
checkInt(sol.leastInterval(["A", "B", "C", "D"], 2), 4, "all unique")
checkInt(sol.leastInterval(["A", "A", "B", "B"], 2), 5, "mix")
print("All tests passed!")

File: problems/09_heap_priority_queue/0621_task_scheduler/solution.dart
// Problem: 0621. Task Scheduler
// Difficulty: Medium

int leastInterval(List<String> tasks, int n) {
  final cnt = <String, int>{};
  for (final t in tasks) cnt[t] = (cnt[t] ?? 0) + 1;
  var maxF = 0;
  for (final v in cnt.values) if (v > maxF) maxF = v;
  var maxCount = 0;
  for (final v in cnt.values) if (v == maxF) maxCount++;
  final formula = (maxF - 1) * (n + 1) + maxCount;
  return tasks.length > formula ? tasks.length : formula;
}

void _expect(int a, int e, String msg) {
  if (a != e) throw StateError('FAIL $msg: $a vs $e');
}

void main() {
  _expect(leastInterval(['A', 'A', 'A', 'B', 'B', 'B'], 2), 8, 'ex1');
  _expect(leastInterval(['A', 'A', 'A', 'B', 'B', 'B'], 0), 6, 'no cool');
  _expect(leastInterval(['A', 'A', 'A', 'A', 'A', 'A', 'B', 'C', 'D', 'E', 'F', 'G'], 2), 16, 'big');
  _expect(leastInterval(['A'], 2), 1, 'one');
  _expect(leastInterval(['A', 'B', 'C', 'D'], 2), 4, 'all unique');
  _expect(leastInterval(['A', 'A', 'B', 'B'], 2), 5, 'mix');
  print('All tests passed!');
}
